src/mlpack/methods/ne/neat.hpp

```diff
@@ -0,0 +1,886 @@
+
+    if (p < crossoverProb) {
+      ssize_t idx1 = mlpack::math::RandInt(0, speciesSize);
+      ssize_t idx2 = mlpack::math::RandInt(0, speciesSize);
+      if (idx1 != idx2) {
+        Crossover(species.aGenomes[idx1], species.aGenomes[idx2], childGenome);
+      } else {
+        return false;
+      }
+    } else {
+      ssize_t idx = mlpack::math::RandInt(0, speciesSize);
+      childGenome = species.aGenomes[idx];
+    }
+
+    Mutate(childGenome);
+
+    return true;
+  }
+
+
+  // Initialize population.
+  void InitPopulation() {
+    aPopulation = Population(aSeedGenome, aPopulationSize);
+  }
+
+  // Reproduce next generation of population.
+  void Reproduce() {
+    // keep best genome.
+    Genome lastBestGenome = aPopulation.BestGenome();
+
+    // Remove weak genomes in each species.
+    CullSpecies(aPopulation, aCullSpeciesPercentage);
+
+    // Remove stale species, weak species.
+    if (aPopulation.aSpecies.size() > 10) {
+      RemoveStaleSpecies(aPopulation);
+      //RemoveWeakSpecies(aPopulation);
+    }
+
+    // Breed children in each species. 
+    std::vector<double> speciesAverageRank;
+    CalcSpeciesAverageRank(aPopulation, speciesAverageRank);
+
+    //DEBUGGING!!!!!!!!!
+    printf("Species average rank are: ");
+    for (ssize_t s=0; s<speciesAverageRank.size(); ++s) {
+      std::cout<< speciesAverageRank[s] << "  ";
+    }
+    printf("\n");
+    //DEBUGGING!!!!!!!!!
+
+    double totalAverageRank = std::accumulate(speciesAverageRank.begin(), speciesAverageRank.end(), 0);
+    std::vector<Genome> childGenomes;
+    for (ssize_t i=0; i<aPopulation.aSpecies.size(); ++i) {
+      // number of child genomes by this species.
+      ssize_t numBreed = std::floor(speciesAverageRank[i] * aPopulationSize / totalAverageRank) - 1;
+      ssize_t numBreedSuccess = 0;
+      while (numBreedSuccess < numBreed) {
+        Genome genome;
+        bool hasBaby = BreedChild(aPopulation.aSpecies[i], genome, aCrossoverRate);
+        if (hasBaby) {
+          childGenomes.push_back(genome);
+          ++numBreedSuccess;
+        }
+      }
+    }
+
+    // Keep the best in each species.
+    CullSpeciesToOne(aPopulation);  // NOTICE: I found that this makes XOR test converges fast.
+
+    // Random choose species and breed child until reach population size.
+    childGenomes.push_back(lastBestGenome);
+    ssize_t currentNumGenome = childGenomes.size() + aPopulation.PopulationSize();
+    while (currentNumGenome < aPopulationSize) {
+      ssize_t speciesIndex = mlpack::math::RandInt(0, aPopulation.aSpecies.size());
+      Genome genome;
+      bool hasBaby = BreedChild(aPopulation.aSpecies[speciesIndex], genome, aCrossoverRate);
+      if (hasBaby) {
+        childGenomes.push_back(genome);
+        ++currentNumGenome;
+      }
+    }
+
+    // Speciate genomes into new species.
+    std::vector<Genome> populationGenomes;
+    AggregateGenomes(aPopulation, populationGenomes);
+    aPopulation.aSpecies.clear();
```

Comment:
If we clear all species, we lose the information about the species age, also I'm not sure, why we have to reassign the genomes.